### PR TITLE
feat: ejection transactor

### PR DESCRIPTION
**Pull request by cody-littley**
## Why are these changes needed?
This PR introduces the ejection transactor, which encapsulates business logic for communicating with onchain ejection contracts.

### CODE REVIEW

File: disperser/controller/metadata/reference_block_provider.go
Comment (cody-littley) [inline disperser/controller/metadata/reference_block_provider.go:1]: This file got duplicated in my other PRs, due to the way I've been splitting smaller PRs off of my feature branch.

File: ejector/ejection_transactor.go [hunk 1]
```diff
@@ -22,3 +31,257 @@ type EjectionTransactor interface {
 	// Complete the ejection proceedings against the operator with the given address.
 	CompleteEjection(ctx context.Context, addressToEject gethcommon.Address) error
 }
+
+var _ EjectionTransactor = &ejectionTransactor{}
+
+// ejectionTransactor is the production implementation of the EjectionTransactor interface.
+//
+// The ejection transactor is thread safe, although parallel calls may result in duplicate work (i.e. two calls might
```
Comment (litt3) [inline ejector/ejection_transactor.go:39]: Here you claim that the ejection transactor _is_ threadsafe, but it uses the periodic rbn provider, which claims to _not_ be threadsafe.

IIRC when reviewing the periodic provider, there was actually something that made it not threadsafe, though I don't recall the specifics.

File: ejector/ejection_transactor.go [hunk 2]
```diff
@@ -22,3 +31,257 @@ type EjectionTransactor interface {
...
+	idToEject, err := e.validatorIDToAddressConverter.ValidatorAddressToID(ctx, addressToEject)
+	if err != nil {
+		return fmt.Errorf("failed to get validator ID from address: %w", err)
+	}
+
+	quorums, err := e.validatorQuorumLookup.GetQuorumsForValidator(ctx, idToEject, rbn)
+	if err != nil {
+		return fmt.Errorf("failed to get quorums for validator: %w", err)
+	}
+	quorumBytes := eth.QuorumListToBytes(quorums)
```
Comment (litt3) [inline ejector/ejection_transactor.go:184]: would it make sense to check if this is empty before submitting the ejection transaction?

File: ejector/ejection_transactor.go [hunk 3]
```diff
@@ -22,3 +31,257 @@ type EjectionTransactor interface {
...
+		Context: ctx,
+		From:    e.selfAddress,
+		Signer:  e.signer,
+	}
+
+	if e.maxGasOverride != 0 {
+		opts.GasLimit = e.maxGasOverride
+	}
+
+	txn, err := e.transactor.CompleteEjection(opts, addressToEject, quorumBytes)
```
Comment (litt3) [inline ejector/ejection_transactor.go:196]: Is it a problem that the transaction could fail to be mined onchain, and this code would never know? What would be the implications of this transaction failing to get mined?

File: ejector/ejection_transactor.go [hunk 4]
```diff
@@ -22,3 +31,257 @@ type EjectionTransactor interface {
...
+
+	// The maximum gas limit to use for transactions.
+	maxGasOverride uint64
+}
+
+// Create a new EjectionTransactor.
+func NewEjectionTransactor(
+	logger logging.Logger,
+	client bind.ContractBackend,
+	ejectionContractAddress gethcommon.Address,
```
Comment (litt3) [inline ejector/ejection_transactor.go:73]: Should you also validate that these aren't 0x0 contract addresses?

File: ejector/ejection_transactor.go [hunk 5]
```diff
@@ -22,3 +31,257 @@ type EjectionTransactor interface {
...
+	registryCoordinatorAddress gethcommon.Address,
+	selfAddress gethcommon.Address,
+	privateKey *ecdsa.PrivateKey,
+	chainID *big.Int,
```
Comment (litt3) [inline ejector/ejection_transactor.go:77]: claude alleges `NewKeyedTransactorWithChainID` will panic if `chainID` is 0

File: ejector/ejection_transactor.go (hunk 1 above)
Comment (cody-littley) [reply ejector/ejection_transactor.go:39]: Added a mutex into the reference block provider to make it thread safe. It's not something I expect to see much contention in, given how we use it. So I think a mutex should be performant enough for now.

File: ejector/ejection_transactor.go (hunk 2 above)
Comment (cody-littley) [reply ejector/ejection_transactor.go:184]:
```
	if len(quorums) == 0 {
		return fmt.Errorf("cannot complete ejection: validator %s is not present in any quorum at RBN %d",
			idToEject.Hex(), rbn)
	}
```

File: ejector/ejection_transactor.go (hunk 3 above)
Comment (cody-littley) [reply ejector/ejection_transactor.go:196]: If the transaction is not mined, then eventually the ejector will attempt to eject the same validator again. As long as the ejector continues retrying, the ejection workflow will complete once the ejection transaction finally gets mined.

Note that the ejection manager is capable of handling ejections that are "mid-phase". That is to say, failure to mine the ejection transaction will not cause the ejector to enter a broken state. As long as transactions eventually start getting mined, the ejector will converge on a good state.

File: ejector/ejection_transactor.go (hunk 4 above)
Comment (cody-littley) [reply ejector/ejection_transactor.go:73]: Added checks. I don't mind defensive checks like this, but I don't always think they are necessary. Specifically for contract addresses, the offchain contract directory code will return an error if the address is a zero address.

File: ejector/ejection_transactor.go (hunk 5 above)
Comment (cody-littley) [reply ejector/ejection_transactor.go:77]: It's not obvious to me when reading the code that this will happen, and if it does happen then that's dumb API. Can you ask claude to explain its reasoning? It's currently tricky for me to test this edge case.

Comment (litt3) [reply ejector/ejection_transactor.go:77]:
```
func newModernSigner(chainID *big.Int, fork forks.Fork) Signer {
	if chainID == nil || chainID.Sign() <= 0 {
		panic(fmt.Sprintf("invalid chainID %v", chainID))
	}
```

I dug down through the call, and found the same thing. This will panic if == 0

I don't know, probably fine to just ignore. Dumb API, but also I'm not sure how a chainId of 0 is ever gonna arrive here

**Review by litt3 — DISMISSED** [review]

File: ejector/ejection_transactor.go [hunk 6]
```diff
@@ -22,3 +31,257 @@ type EjectionTransactor interface {
...
+	if err != nil {
+		return false, fmt.Errorf("failed to get quorums for validator: %w", err)
+	}
+
+	return len(quorums) > 0, nil
+}
+
+func (e *ejectionTransactor) StartEjection(
+	ctx context.Context,
+	addressToEject gethcommon.Address) error {
```
[the remark this answers is not in the document]
Comment (cody-littley) [reply ejector/ejection_transactor.go]: fixed

File: ejector/ejection_transactor.go (hunk 5 above)
Comment (cody-littley) [reply ejector/ejection_transactor.go:77]: added check

```
	if chainID.Sign() <= 0 {
		return nil, fmt.Errorf("invalid chainID: %s", chainID.String())
	}
```

Comment (cody-littley) [reply ejector/ejection_transactor.go:77]: So frustrating. `NewKeyedTransactorWithChainID()` returns an error already, whoever write this was just lazy.
